### DISCUSSION TITLE
Question for DF8OE about bootloader

### DISCUSSION THREAD

**Discussion by BitLoose** [df8oe/UHSDR#1935] [Q&A]
Hi Andreas, it was suggested in the McHF group that you are expert on the bootloader and could help me.
I´ve installed the firmware but I am not sure if the bootloader was also installed, as reset is not working with the vtable at 0x08010000.
I can fix the debugger reset address and that then boots.
How can I check if the bootloader, presuming there must be a custom one for McHF, is loaded?
Sorry for the NOOB question, but am new to the project.
Vielen dank! 73, Rob

**Comment by df8oe** [comment]
Hi Rob,
putting a mcHf or an OVI40 to a working state are two steps:
1) load bootlaoder via dfu or STLink
2) load firmware via bootloader and USB-Stick, dfu or STLink
Both steps are mandatory and you MUST proceed them. Of course the bootloader must match your hardware.

Resetvector of STM points to bootloader. Main benefit of bootloader is that it can handle firmware updates from now on with a simple USB-stick where the firmware binary resides. Without bootloader resetvector points to nowhereland.

**Reply by BitLoose** [reply]
That makes so much sense!  So, for a 32F407VG based setup, is there a bootloader on GitHub I should use?
I thought it would be included with the UHSDR download but haven't seen it yet.

**Reply by BitLoose** [reply]
The bootloader doesn't not show in the STMCubeIDE folders from importing UHSDR project, but I can see it using file explorer.
I can import it, but there is no build information, so the question is what am I missing?

**Reply by BitLoose** [reply]
Andreas, as I need to compile the bootloader I´ve been looking for guidance documentation on building the boot loader, in the code there is a define for BOOTLOADER_BUILD

I added it to the preprocessor list but that gives 5 errors because the bootloader files were not imported and there are undefined function references.  I've now added the files.

Is this the correct way to get the bootloaded compiled?

  /* USER CODE BEGIN 1 */
#ifdef BOOTLOADER_BUILD
    Bootloader_CheckAndGoForBootTarget();
  //  we need to do this as early as possible
#endif
  /* USER CODE END 1 */

**Comment by db4ple** [comment]
The bootloader is built from the same source. See makefiles in the mchf-eclipse dir or use the recommended Eclipse environment for building your bootloader (see instructions in Wiki) if that is your thing. 
To build the bootloader you have to do more changes than just setting the define, it also requires using different linker scripts / linker settings in order to have the load address placed where the STM32 needs it during startup.

However, as @df8oe suggest, just flashing the precompiled bootloader should do the job just fine.

**Reply by BitLoose** [reply]
Thanks very much for your help.  Such a pain this, hard to believe it would be so difficult to get the bootloader to compile in STM32CubeIDE?!
I guess I can install the Eclipse IDE,  but seems crazy that all these tools are so incompatible after all these years.

**Reply by df8oe** [reply]
Crazy - but reality. We have choosen an Open Source solution which is available on all operating systems. So our choice was Eclipse - and using "plain Makefile". Debugging in Eclipse works well. CubeIDE is not supported by us and it needs many tuning. It is the personally decision of everyone to step other ways - but if so they do it without support from us.

**Comment by BitLoose** [comment]
I understand, I've been going through the bootloader code, it launches the radio app at 0x08010000 but how do I set the app so that it starts at that address?Currently it's at 0x08007df and varies with each build change.I changed the section address in the script from "= ." to = 0x08010000, no error but also no change.Maybe you know!On 12 Nov 2022 08:27, DF8OE ***@***.***> wrote:
Crazy - but reality. We have choosen an Open Source solution which is available on all operating systems. So our choice was Eclipse - and using "plain Makefile". Debugging in Eclipse works well. CubeIDE is not supported by us and it needs many tuning. It is the personally decision of everyone to step other ways - but if so they do it without support from us.
